Add CRL processing KeyUsage validation.

cpu opened the pull request:
## Description

This branch extends the CRL validation code that was added in xxx to handle the key usage (KU) validation described by [RFC 5280  §6.3.3](https://www.rfc-editor.org/rfc/rfc5280#section-6.3.3):
> If a key usage extension is present in the CRL issuer's certificate, verify that the cRLSign bit  is set. 

Importantly, we do this only for **intermediate certificate authorities**, and not for trust anchors that have issued a CRL. Since we parse intermediate certificates from DER we can easily grab the key usage extension (if present) without any ecosystem breaking changes. Trust anchors on the other hand are expressed minimally, and provided by external libraries like [webpki-roots](https://github.com/rustls/webpki-roots) or [rustls-native-certs](https://github.com/rustls/rustls-native-certs). Adding a Key Usage field to trust anchors for this purpose feels like a disproportionate breaking change for limited benefit and should be considered separately.

The individual commits to achieve the goal of key usage validation for CRL issuers are described in detail below:

### tests: update generate.py to include CRL KU tests.
This commit updates the `generate.py` `client_auth_revocation` test suite to include a series of tests that exercise key usage validation when processing CRLs.

Since we can't yet expect specific errors it's tricky to demonstrate key usage validation is happening as expected. To work around this we include one test case (`_ee_not_revoked_wrong_ku_ee_depth`) that will only pass when key usage is not being enforced. Once key usage is enforced, it will fail and require updating.

### tests: generate client_auth_revocation tests.

Created by running:
```
cd tests
./generate.py --no-tls-server-certs --no-signatures --no-clientauth
```

### cert: remember optional KU certificate extension bitstring.
This commit updates `remember_cert_extension` to capture the `id-ce-keyUsage` extension, if present, into the `cert.key_usage` field. Some care is required here since unlike all of the other remembered certificate extensions the KeyUsage is not wrapped in a `Sequence` but is a stand-alone `BitString`.

The comment that was on the `id-ce-keyUsage` field match arm in `remember_cert_extension` about ignoring the key usage extension for the purpose of certificate validation is lifted into `verify_cert.rs`'s `check_issuer_independent_properties` where it was previously referenced. We will continue to ignore KU for the purposes of certificate validation and only intend to use this field when validating CRLs.

### der: add support for ASN.1 bit string flags.

This commit updates `der.rs` to add support for parsing and working with ASN.1 BitStrings that represent flags. A `BitStringFlags` type can be constructed from raw DER using `bit_string_flags`. Once in-hand, specific flag values can be tested using `BitStringFlags.bit_set`.

At the stage of this commit, no code uses this functionality so we allow `unused` to let this commit build independently. The next step will be using this code when processing CRL key usage flags and will remove these allows.

### verify_cert: enforce CRL issuer KeyUsage.

This commit updates the `verify_cert.rs` processing of CRLs to enforce that if an intermediate certificate has issued a CRL, and the intermediate cert specifies a KeyUsage extension, that the cRLSign bit of the KeyUsage is included. If not, we return an error indicating this.

If no KeyUsage extension is included then we follow the conventional behaviour of web pki validators and assume the absence means "any key usage", including CRL issuance.

At present we only track/validate KeyUsage for intermediate CA certificates and not for trust anchors. It will not be difficult to
extend this processing to trust anchors but introducing a new field on the trust anchor type needs to be carefully considered with respect to downstream ramifications on crates like webpki-roots and the platform verifier crate.

**Note: The "canary" test we added earlier now fails as expected. To keep the commit building cleanly we comment out the generated test case in this commit, and will restore it in a subsequent commit to avoid the large generated test diff being rolled into this commit.**

<details>
<summary>Checking out the branch at this commit and uncommenting the  `'ee_not_revoked_wrong_ku_ee_depth` test cases produces the expected failure:</summary>

```
---- ee_not_revoked_wrong_ku_ee_depth stdout ----
thread 'ee_not_revoked_wrong_ku_ee_depth' panicked at 'assertion failed: `(left == right)`
  left: `Err(UnknownIssuer)`,
 right: `Ok(())`', tests/client_auth_revocation.rs:180:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    ee_not_revoked_wrong_ku_ee_depth

test result: FAILED. 16 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```
</details>

## tests: update expected error for incorrect KU test.

Previously in this branch we added a test case that would explicitly only pass when CRL key usage was not being enforced. Now that it _is_ being enforced, this commit changes the expected error so that the regenerated test will pass.

## tests: regenerate client_auth_revocation tests.

To pick up the change in the previous commit, we regenerate the test suite.

Now the `ee_not_revoked_wrong_ku_ee_depth` test passes again. :tada:

Resolves https://github.com/rustls/webpki/issues/52, updates https://github.com/rustls/webpki/issues/57

cpu commented on src/verify_cert.rs:
```diff
@@ -324,6 +405,39 @@ fn check_eku(
 
+type KeyUsageBit = usize;
+
+// https://www.rfc-editor.org/rfc/rfc5280#section-4.2.1.3
+// const KU_DIGITAL_SIGNATURE: KeyUsageBit = 0;
+// const KU_CONTENT_COMMITMENT: KeyUsageBit = 1;
+// const KU_KEY_ENCIPHERMENT: KeyUsageBit = 2;
+// const KU_DATA_ENCIPHERMENT: KeyUsageBit = 3;
+// const KU_KEY_AGREEMENT: KeyUsageBit = 4;
+// const KU_KEY_CERT_SIGN: KeyUsageBit = 5;
+const KU_CRL_SIGN: KeyUsageBit = 6;
+// const KU_ENCIPHER_ONLY: KeyUsageBit = 7;
+// const KU_DECIPHER_ONLY: KeyUsageBit = 8;
```
I personally thought it was nicer to have the unused bits represented here for context but if folks disagree I can remove the commented out bits, leaving only `KU_CRL_SIGN`.

ctz commented:
> Test coverage is already provided by the existing CRL test suite from
> the pt2 PR (though it's difficult to tell since all test cases assert
> UnknownIssuer until the error specificity from path building is
> addressed).

Would be ideal if there were a test case that went from failing to passing here -- ie, issuer had a KU extension which had only (say) keyCertSign, plus a CRL that doesn't revoke the EE cert. That would mean the test previously accepted the certificate (having processed the CRL and found EE not in it) and then stopped accepting it after this PR.

djc commented on src/verify_cert.rs:
```diff
@@ -399,6 +405,39 @@ fn check_eku(
+// const KU_DECIPHER_ONLY: KeyUsageBit = 8;
+
+// https://www.rfc-editor.org/rfc/rfc5280#section-4.2.1.3
+fn check_ku(
```
I'm on the fence about the proliferation of the `ku` acronym. I think I would prefer to expand to `check_key_usage()` here (and maybe even for the field?).

djc commented on src/verify_cert.rs:
```diff
@@ -399,6 +405,39 @@ fn check_eku(
     }
 }
 
+type KeyUsageBit = usize;
```
I tend to dislike type aliases for this type of thing: they obfuscate but don't provide type safety. In this case, I think a translation to an enum would be more idiomatic?

```rust
#[repr(u8)]
enum KeyUsageMode {
    // DigitalSignature = 0,
    // ContentCommitment = 1,
    CrlSign = 6,
    ..
}
```

cpu replied:
Sounds good, I think using `#[repr(u8)]` was the missing piece that kept me from thinking of this initially.

cpu replied:
Updated to remove the type alias in place of an enum. Thanks!

djc commented on src/verify_cert.rs:
```diff
@@ -399,6 +405,39 @@ fn check_eku(
+    input: Option<untrusted::Input>,
+    required_ku_bit_if_present: KeyUsageBit,
+) -> Result<(), Error> {
+    match input {
```
I would linearize this (and spell `bit_string` with an underscore):

```rust
let Some(bit_string) = input else {
    return Ok(());
};
```

cpu replied:
Done.

cpu commented on src/cert.rs:
```diff
@@ -52,6 +52,9 @@ pub struct Cert<'a> {
     /// raw basic constraints extension (if any).
     pub basic_constraints: Option<untrusted::Input<'a>>,
 
+    /// key usage extension (if any).
```
Done (though I think it might be less useful now that the field isn't public based on the updates in https://github.com/rustls/webpki/pull/66)

cpu commented:
> cpu force-pushed the cpu-crl-support-pt3-ku branch from 4dc95e5 to 62cb079 

Rebased with the changes in https://github.com/rustls/webpki/pull/66 and addressed djc's feedback. 

I will look at the test coverage bit next.

cpu commented:
> cpu force-pushed the cpu-crl-support-pt3-ku branch from 62cb079 to ca62f02 

Updated to bring the KU-specific test cases from https://github.com/rustls/webpki/pull/66 into this branch.

I also took the approach @ctz recommended and added a canary test that fails if you check out the branch at [83a8e61](https://github.com/rustls/webpki/pull/82/commits/83a8e6124f9f1d213e7588eb96a0280bfed8130d) and run the test suite. It's fixed in subsequent commits.

The PR description is also updated to describe the new commits and to show the failure from the test output in the intermediate stage.

djc commented:
I think it's good to add a test that is verified to fail before fixing it, but ideally we shouldn't keep the commit history like that when merging -- it's useful when the main branch history always compiles and passes tests (as far as possible) for bisection.

djc commented on src/der.rs:
Style: `bitstring` -> `bit_string`? (Or maybe just `bits`?)

djc commented on src/der.rs:
Style: suggest writing this as

```rust
match padding_bits < 0 && (last_byte & padding_mask) != 0 {
    true => Err(Error::BadDer),
    false => Ok(BitStringFlags { raw_bits }),
}
```

cpu replied:
Fixed up. Thanks.

djc commented on tests/client_auth_revocation.rs:
```diff
@@ -179,7 +179,7 @@ fn ee_not_revoked_wrong_ku_ee_depth() {
     assert_eq!(
         check_cert(ee, intermediates, ca, RevocationCheckDepth::EndEntity, crls),
-        Ok(())
+        Err(webpki::Error::UnknownIssuer)
```
This means that these tests failed in the previous commit?

cpu replied:
Yup :ballot_box_with_check:

djc commented:
I'd probably do it before review, but I don't care that much.

cpu commented:
@ctz @djc This branch should be ready for review now. I've rebased it with the work that landed prior, integrated Djc's feedback, and update the pull request description.

Notably each commit builds + tests cleanly per djc's request. Ctz's idea of having a test that fails when enforcement is enabled is still present, but now requires a little bit of leg work to demonstrate (_or you can look at the output I included in the pull request description if you trust me :-)_) - You'll need to check out the branch at 3c5287a0ec5d0e1d589306409bcdb844f229aff4 and then uncomment the `ee_not_revoked_wrong_ku_ee_depth` test in `tests/client_auth_revocation.rs`. Running `cargo test` will then fail as expected. The commits after 3c5287a0ec5d0e1d589306409bcdb844f229aff4 update the test generator for the correct expected error with enforcement enabled, and regenerate the tests to restore `ee_not_revoked_wrong_ku_ee_depth` in a passing state. I hope that's a good balance between the request to definitively show enforcement happens via test failure -> test pass while also keeping the entire branch happy for bisecting.

ctz commented:
> I think it's good to add a test that is verified to fail before fixing it, but ideally we shouldn't keep the commit history like that when merging -- it's useful when the main branch history always compiles and passes tests (as far as possible) for bisection.

There are some other options here -- having the test but marking it `#[ignore]` for a limited period (preferable), or having _and_ running the test but marking it `#[should_panic]` (noisy, less good).

ctz reviewed (approved):
looking good!

djc commented:
Oh yeah, marking the test `#[ignore]` seems like a fine way to do it. Though generally I'll trust a serious PR author who asserts that they've verified the test failed before the fix.

djc commented on tests/client_auth_revocation.rs:
```diff
@@ -14,8 +14,6 @@
 
 extern crate webpki;
 
-use webpki::EndEntityOrCa;
```
This is not actually from clippy, but from the compiler lints. Maybe add some gunk to the CI workflow to make sure it denies warnings?

cpu replied:
Ah good catch. I'll look at that in a follow-up.

cpu replied:
https://github.com/rustls/webpki/pull/86

cpu commented:
Thanks for the reviews! I'll prefer `#[ignore]` if I run into this situation again.
